### DISCUSSION TITLE
Controlling nodes for ceph

### DISCUSSION THREAD

**Discussion by bothorsen** [rook/rook#10240] [General]
I'm trying to find the right way to choose the nodes that will have ceph storage, and could use a bit of help with this. The thing is that it's very hard to figure out the consequences of my choices, based on the current documentation.

My small cluster has three control planes and four worker nodes. Each control plane has one disk, worker nodes has two. This is the current and simple setup, but as it grows, it will likely become less simple.

I'd like to run ceph on the worker nodes only and use the entire sdb for the nodes. So here are the questions I have:

1. Should I just use sdb or should I use lvm for the ceph disks? I don't think that I'll need more than one disk, but maybe that's naive.
2. If I just set it up to use all nodes, is this worker nodes only or really all nodes? If not, what's the best way to say that it should only do the worker nodes?
3. Is it wrong to think that only worker nodes should be used? I keep my control planes untainted, but I'm not sure how to think about this.
4. If I set it up to use sdb on all nodes, will it give errors for those nodes that only have one disk or be fine and just use those where it's present?

Sometimes when I answer development questions, I can see that the questions are wrong because the underlying thinking is wrong. If this is the case here, please let me know. You don't have to answer the questions directly, as I just need a solution that fits what I would like to achieve.

Thank you for any help you can give me.

Bo Thorsen.

**Comment by travisn** [comment]
> I'd like to run ceph on the worker nodes only and use the entire sdb for the nodes. So here are the questions I have:
> 
> 1. Should I just use sdb or should I use lvm for the ceph disks? I don't think that I'll need more than one disk, but maybe that's naive.

You can set the `deviceFilter: sdb` (with `useAllDevices: true`) if you only want that device to be used on each node. If you decide to use both/all disks in the future you could remove the filter for all devices, or update the filter.

> 2. If I just set it up to use all nodes, is this worker nodes only or really all nodes? If not, what's the best way to say that it should only do the worker nodes?

If your control plane nodes are not tainted, you can use the [placement.nodeAffinity](https://github.com/rook/rook/blob/release-1.9/deploy/examples/cluster.yaml#L146-L155). In that example, rook will only start ceph pods on nodes with the `role=storage-node` label. 

> 3. Is it wrong to think that only worker nodes should be used? I keep my control planes untainted, but I'm not sure how to think about this.

It is common to reserve the control plane nodes for K8s and not run rook there, but they can also be used if needed, for example in small clusters where there just aren't enough nodes. One possibility is also to set the mons to run on the control nodes and the OSDs and other daemons on the workers. The mons are conceptually the control plane for ceph. There is specific `placement` for mons if you want to go this route, instead of just using `placement.all`.

> 4. If I set it up to use sdb on all nodes, will it give errors for those nodes that only have one disk or be fine and just use those where it's present?

If a node doesn't have an `sdb` device available, the node simply won't have any OSDs created (no errors).

> Sometimes when I answer development questions, I can see that the questions are wrong because the underlying thinking is wrong. If this is the case here, please let me know. You don't have to answer the questions directly, as I just need a solution that fits what I would like to achieve.

No worries on the questions! Ramping up is certainly not simple.
